### PR TITLE
DAOS_10172 rebuild: wait dtx resync globally

### PR DESCRIPTION
**Pull request by wangdi1**
Let's wait for dtx resync to finish on all engines, then
start rebuild, to make sure rebuild will not miss any
un-commit data during rebuild/reintegration.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Review by Nasf-Fan — APPROVED** [review]

**Comment by wangdi1** [comment]
failed due to DAOS-10051. retriever the test.

File: src/rebuild/scan.c [hunk 1]
```diff
@@ -811,12 +811,12 @@ rebuild_scan_leader(void *data)
...
 
-	D_DEBUG(DB_REBUILD, DF_UUID "check resync %u < %u\n",
+	D_DEBUG(DB_REBUILD, DF_UUID "check resync %u/%u < %u\n",
 		DP_UUID(rpt->rt_pool_uuid), rpt->rt_pool->sp_dtx_resync_version,
-		rpt->rt_rebuild_ver);
+		rpt->rt_global_dtx_resync_version, rpt->rt_rebuild_ver);
 
 	/* Wait for dtx resync to finish */
-	while (rpt->rt_pool->sp_dtx_resync_version < rpt->rt_rebuild_ver)
+	while (rpt->rt_global_dtx_resync_version < rpt->rt_rebuild_ver)
```
Comment (liuxuezhao) [inline src/rebuild/scan.c:819]: adding a "dss_sleep(5);" here probably a little better?

**Review by liuxuezhao — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
sigh, need increase the timeout for rebuild test, since this patch will need the rebuild to wait all nodes to finish dtx resync.

**Comment by johannlombardi** [comment]
Please inspect. This is a blocker for master to reopen.

File: src/rebuild/scan.c [hunk 2]
```diff
@@ -812,13 +812,19 @@ rebuild_scan_leader(void *data)
...
 	/* Wait for dtx resync to finish */
-	while (rpt->rt_pool->sp_dtx_resync_version < rpt->rt_rebuild_ver)
-		ABT_thread_yield();
+	while (rpt->rt_global_dtx_resync_version < rpt->rt_rebuild_ver) {
+		if (rpt->rt_abort || rpt->rt_finishing) {
+			D_INFO("shutdown rebuild "DF_UUID": "DF_RC"\n",
+			       DP_UUID(rpt->rt_pool_uuid), DP_RC(-DER_SHUTDOWN));
+			D_GOTO(out, rc = -DER_SHUTDOWN);
+		}
+		dss_sleep(2 * 1000);
```
Comment (gnailzenh) [inline src/rebuild/scan.c:826]: better to add a debug message and show the total waiting time, otherwise we might have no idea when things go south.

**Review by gnailzenh — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

File: src/rebuild/scan.c (hunk 2 above)
Comment (wangdi1) [reply src/rebuild/scan.c:826]: yeah,  I can add this global dtx resync version to rebuild console message.

**Pull request merged**